drivers/tty/serial/8250/8250_ni.c

```diff
@@ -0,0 +1,461 @@
+#include <linux/property.h>
+
+#include "8250.h"
+
+/* TFS - TX FIFO Size */
+#define NI16550_TFS_OFFSET	0x0C
+/* RFS - RX FIFO Size */
+#define NI16550_RFS_OFFSET	0x0D
+
+/* PMR - Port Mode Register */
+#define NI16550_PMR_OFFSET	0x0E
+/* PMR[1:0] - Port Capabilities */
+#define NI16550_PMR_CAP_MASK			0x03
+#define NI16550_PMR_NOT_IMPL			0x00 /* not implemented */
+#define NI16550_PMR_CAP_RS232			0x01 /* RS-232 capable */
+#define NI16550_PMR_CAP_RS485			0x02 /* RS-485 capable */
+#define NI16550_PMR_CAP_DUAL			0x03 /* dual-port */
+/* PMR[4] - Interface Mode */
+#define NI16550_PMR_MODE_MASK			0x10
+#define NI16550_PMR_MODE_RS232			0x00 /* currently 232 */
+#define NI16550_PMR_MODE_RS485			0x10 /* currently 485 */
+
+/* PCR - Port Control Register */
+#define NI16550_PCR_OFFSET	0x0F
+#define NI16550_PCR_RS422			0x00
+#define NI16550_PCR_ECHO_RS485			0x01
+#define NI16550_PCR_DTR_RS485			0x02
+#define NI16550_PCR_AUTO_RS485			0x03
+#define NI16550_PCR_WIRE_MODE_MASK		0x03
+#define NI16550_PCR_TXVR_ENABLE_BIT		(1 << 3)
+#define NI16550_PCR_RS485_TERMINATION_BIT	(1 << 6)
+
+/* flags for ni16550_device_info */
+#define NI_HAS_PMR		0x0001
+
+struct ni16550_device_info {
+	unsigned int uartclk;
+	uint8_t prescaler;
+	unsigned int flags;
+};
+
+struct ni16550_data {
+	int line;
+};
+
+static int ni16550_enable_transceivers(struct uart_port *port)
+{
+	uint8_t pcr;
+
+	pcr = port->serial_in(port, NI16550_PCR_OFFSET);
+	pcr |= NI16550_PCR_TXVR_ENABLE_BIT;
+	dev_dbg(port->dev, "enable transceivers: write pcr: 0x%02x\n", pcr);
+	port->serial_out(port, NI16550_PCR_OFFSET, pcr);
+
+	return 0;
+}
+
+static int ni16550_disable_transceivers(struct uart_port *port)
+{
+	uint8_t pcr;
+
+	pcr = port->serial_in(port, NI16550_PCR_OFFSET);
+	pcr &= ~NI16550_PCR_TXVR_ENABLE_BIT;
+	dev_dbg(port->dev, "disable transceivers: write pcr: 0x%02x\n", pcr);
+	port->serial_out(port, NI16550_PCR_OFFSET, pcr);
+
+	return 0;
+}
+
+static int ni16550_config_rs485(struct uart_port *port,
+				struct ktermios *termios,
+				struct serial_rs485 *rs485)
+{
+	uint8_t pcr;
+
+	/* "rs485" should be given to us non-NULL. */
+	if (WARN_ON(rs485 == NULL))
+		return -EINVAL;
+
+	pcr = port->serial_in(port, NI16550_PCR_OFFSET);
+	pcr &= ~NI16550_PCR_WIRE_MODE_MASK;
+
+	if (rs485->flags & SER_RS485_ENABLED) {
+		/* RS-485 */
+		if ((rs485->flags & SER_RS485_RX_DURING_TX) &&
+		    (rs485->flags & SER_RS485_RTS_ON_SEND)) {
+			dev_dbg(port->dev, "Invalid 2-wire mode\n");
+			return -EINVAL;
+		}
+
+		if (rs485->flags & SER_RS485_RX_DURING_TX) {
+			/* Echo */
+			dev_vdbg(port->dev, "2-wire DTR with echo\n");
+			pcr |= NI16550_PCR_ECHO_RS485;
+		} else {
+			/* Auto or DTR */
+			if (rs485->flags & SER_RS485_RTS_ON_SEND) {
+				/* Auto */
+				dev_vdbg(port->dev, "2-wire Auto\n");
+				pcr |= NI16550_PCR_AUTO_RS485;
+			} else {
+				/* DTR-controlled */
+				/* No Echo */
+				dev_vdbg(port->dev, "2-wire DTR no echo\n");
+				pcr |= NI16550_PCR_DTR_RS485;
+			}
+		}
+	} else {
```

Comment:
Recent upstream changes in 63f75fea3a72f93a022d2f027842fe56dd0ebb08 have revealed some misalignment between the NI interpretation of these and what they are meaning for the rest of the kernel. Some of that probably stems from our original implementation pre-dating the serial_rs485 structure.

The way the Linux kernel has currently defined RS-485 support, only "4-Wire / RS-422" and "2-Wire Auto" really make sense. Setting `SER_RS485_ENABLED` implies that you want the driver to handle enabling/disabling the transceiver for you, and the other flags only alter the methods used for automatic control.

We had interpreted this differently, with `SER_RS485_ENABLED` meaning that you were half duplex, but allowed for both automatic and manual control modes.

I think this should be reworked to align with the upstream interpretation. This means eliminating the "2-wire DTR no echo" and "2-wire DTR with echo" implementations. Those modes have always been hidden in NI-VISA on NI Linux RT, so I don't anticipate there are many users of those modes, if any at all.

I have some other documentation written up as we've looked at this for other drivers. Reach out to me for more details.

Reply:
I thought the ni16550 UART would require writing to the ACR register to enable automatic transceiver control, in addition to the PCR. I don't see that here, nor do I see it in the 8250_port driver. This is something to dive into more to understand how this is (or isn't) working.

Reply:
The old implementation never did anything with `ACR` before  (https://github.com/ni/linux/blob/nilrt/master/5.15/drivers/tty/serial/8250/8250_ni16550.c#L9) but I did notice that when we tried with the PCI/PXI ones in fdc2de87124f we do write to `ACR` there.

Reply:
> The old implementation never did anything with ACR

OK, looking through some of the implementation as stirred up some old memories. Prior to existing in the NI16550 form, the PCR was implemented in external logic. It was necessary to program both the ACR (for automatic DTR control) and PCR (to configure external logic) to get the proper automatic transceiver control.

In the NI16550 it is all integrated, so there are explicit RxEnable and TxEnable lines/signals. This allows for automatic transceiver control without explicitly configuring the UART's Auto-DTR through the ACR register.

So, not programming the ACR should be fine where this is used. If it is ever applied to PCI-based boards then more work will be needed to configure the Auto-DTR behavior.

Reply:
I think this is fine for now, with the understanding that we'll need to revisit the RS-485 mode configuration.

Reply:
While I was reducing the cases down, I also I went looking at adding ACR programming (since, skimming through the HDL, it _looks_ like it should be harmless to do on these instantiations and makes us ready for the PCI ones?)... and I think I figured out why fdc2de87124f didn't work (acr |= NI16550_ACR_DTR_MANUAL_DTR doesn't work when that constant is zero, so once you acr |= NI16550_ACR_DTR_AUTO_DTR'd you were stuck that way, oops)